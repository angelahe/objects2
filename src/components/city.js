class city {

  constructor(name, latitude, longitude, population) {
    this.Name = name;
    this.Latitude = latitude;
    this.Longitude = longitude;
    this.Population = population;
  }

  show() {
    const attributeString = `${this.Name} ${this.Latitude} ${this.Longitude} ${this.Population}`;
    return attributeString;
  }

  movedIn(numPeople) {
    const populationIn = numPeople;
    return this.Population += populationIn;
  }

  movedOut(numPeople) {
    const populationOut = numPeople;
    return this.Population -= populationOut;
  }

  howBig() {
    switch (true) {
      case (this.Population > 100000) :
        return "City";
        break;
      case (this.Population > 20000) :
        return "Large town";
        break;
      case (this.Population >1000) :
        return "Town";
        break;
      case (this.Population > 100) :
        return "Village";
        break;
      case(this.Population > 0):
        return "Hamlet";
        break;
        default:
          return "Not sure how big this place is...";
    }
  }



}

export default city