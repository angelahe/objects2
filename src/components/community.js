import city from './city';

class community {

  constructor() {
    this.Communities = [];
  }

  addCommunity(name, latitude, longitude, population) {

    const newCommunity = new city(name, latitude, longitude, population);
    this.Communities.push(newCommunity);

  }

  whichSphere(index) {

    if (this.Communities[index].Latitude.includes("N")) {
      return ("Northern Hemisphere");
    } else if (this.Communities[index].Latitude.includes("S")) {
      return ("Southern Hemisphere");
    } else {
      return ("Not of this earth");
    }

  }


  // assume Latitude is stored as a string e.g. "51.0486 N"
  getMostNorthern() {

    let mostNorthern = "";
    let mostNorthLatitude = 0;
    let nextNorthLatitude = 0;

    for(let i=0; i<this.Communities.length; i++ ) {
      if (this.Communities[i].Latitude.includes("N")) {
        nextNorthLatitude = Number(this.Communities[i].Latitude.slice(0, -2));
      }
      if (nextNorthLatitude > mostNorthLatitude) {
        mostNorthLatitude = nextNorthLatitude;
        mostNorthern = this.Communities[i].Name;
      }
    }

    return(mostNorthern);

  }

  getMostSouthern() {
    let mostSouthern = "";
    let mostSouthLatitude = 0;
    let nextSouthLatitude = 0;

    for(let i=0; i<this.Communities.length; i++ ) {
      if (this.Communities[i].Latitude.includes("S")) {
        nextSouthLatitude = Number(this.Communities[i].Latitude.slice(0, -2));
      }
      if (nextSouthLatitude > mostSouthLatitude) {
        mostSouthLatitude = nextSouthLatitude;
        mostSouthern = this.Communities[i].Name;
      }
    }

    return(mostSouthern);
  }

  getPopulation() {
    //let population = 8237550;

    const totalPopulation = this.Communities.reduce(function (acc, obj) {return acc + obj.Population; }, 0);

    return(totalPopulation);
  }
}

export default community;