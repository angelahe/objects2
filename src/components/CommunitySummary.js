import React from 'react';
import '../styles/Styles140.css';
import './community';

class CommunitySummary extends React.Component {
  constructor(props) {
    super(props);
    this.state = this.computeAggregates(props.communities);
  }

  componentWillReceiveProps(nextProps, nextContext) {
    this.setState(this.computeAggregates(nextProps.communities));
  }

  computeAggregates(communities) {
    let mostNorthern = "";
    let mostSouthern = "";
    let totalPopulation = 0;

    mostNorthern = communities.getMostNorthern();
    mostSouthern = communities.getMostSouthern();
    totalPopulation = communities.getPopulation();

    return {
      mostNorthern : mostNorthern,
      mostSouthern : mostSouthern,
      totalPopulation : totalPopulation
    };
  }

  render() {
    return (
      <div className="SummaryText">
        <h2>Community Summary</h2>
        <span>Most Northern: </span>
        <span>{this.state.mostNorthern}</span><br/>
        <span>Most Southern: </span>
        <span>{this.state.mostSouthern}</span><br/>
        <span>Total Population:</span>
        <span>{this.state.totalPopulation}</span><br/>
        <br/>
      </div>
    )
  }
}
export default CommunitySummary