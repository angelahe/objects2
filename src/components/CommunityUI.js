import React from 'react';
import community from './community';
import CityCreate from './CityCreate';
import CityEdit from './CityEdit';
import OneCommunity from './OneCommunity';
import CommunitySummary from './CommunitySummary';
import addbtn from '../images/add_FFFFFF.png';

import '../styles/Styles140.css';

class CommunityUI extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      CommunityList: new community(),
      currentCity: null,
      addShow: false,
      editShow: false
    }
  }

  onBtnCloseClick = (closeType) => {
    if (closeType === "create") this.setState ({addShow: false});
    if (closeType === "edit") this.setState ({editShow: false});
  }

  onBtnAddClick = () => {
    this.setState({addShow: true});
  }

  onBtnEditClick = () => {
    this.setState({editShow: true});
  }

  onCityUpdate = (updated) => {
    console.log("in city update");
    this.state.CommunityList.updateCommunity(updated);

    this.setState({
      CommunityList: this.state.CommunityList,
      editShow: false
    })

  }


  onCityCreate = (name, latitude, longitude, population) => {
    this.state.CommunityList.addCommunity(name, latitude, longitude, population);

    this.setState({
      CommunityList: this.state.CommunityList,
      addShow: false
      }
    )

  }

  render() {
    const communities = this.state.CommunityList.Communities.map((city, index) =>
    <OneCommunity key = {city.Name}
                  city = {city}
    />
    );

    return(
      <div>

        <h1>This is the Community UI</h1>
        <div className="AppContainer">
          <CommunitySummary communities={this.state.CommunityList}/>
          <div className="AppPanel">
            <div className="ItemBox AppHeader">
              <span className="AddItem">Add City</span>
              <button className="AppBtn" onClick={this.onBtnAddClick}>
                <img className="btnImg" src={addbtn} alt="Add"/>
              </button>
            </div>
            <div className = "AppList">
              {communities}
            </div>
          </div>
          <div className = "AppPanel">
            {(this.state.addShow)
              ? <CityCreate createClicked = {this.onCityCreate}
                            closeClicked={this.onBtnCloseClick}
                />
              :null
            }
            {(this.state.editShow)
              ? <CityEdit
                         city={this.state.currentCity}
                         updateCity = {this.onCityUpdate}
                         closeClicked={this.onBtnCloseClick}
                />
              : null
              }
          </div>

        </div>
      </div>
    )
  }
}
export default CommunityUI
