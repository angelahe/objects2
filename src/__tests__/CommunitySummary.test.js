import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import CommunitySummary from "../components/CommunitySummary";
import CommunityUI from "../components/CommunityUI";
import community from "../components/community";
import city from "../components/city";
import renderer from "react-test-renderer";

const props = new CommunityUI();

it('renders the mostNorthern, mostSouthern, and total population', () => {
  let communityList = new community();

  communityList.addCommunity("Calgary", "51.0486 N", "114.0708 W", 1239000);
  communityList.addCommunity("Sydney", "33.8688 S", "151.2093 E", 4452000);
  communityList.addCommunity("Edmonton", "53.5444 N", "113.4909 W", 932550);
  communityList.addCommunity("Auckland", "36.8485 S", "174.7633 E", 1614000);

  const div = document.createElement('div');
  const component = ReactDOM.render(<CommunitySummary communities={communityList}/>, div);
  expect(component.state.mostNorthern).toEqual("Edmonton");
  expect(component.state.mostSouthern).toEqual("Auckland");
  expect(component.state.totalPopulation).toEqual(8237550);

});

it('renders a summary with no cities', () => {
  const wrapper = shallow(<CommunitySummary communities={new community()} />);

  expect(wrapper.state('totalPopulation')).toEqual(0);
});

test('CommunitySummary renders correctly', () => {
  const tree = renderer.create(<CommunitySummary communities={props.state.CommunityList} />).toJSON();
  expect(tree).toMatchSnapshot();
});