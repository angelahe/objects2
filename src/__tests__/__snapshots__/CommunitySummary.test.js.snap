// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommunitySummary renders correctly 1`] = `
<div
  className="SummaryText"
>
  <h2>
    Community Summary
  </h2>
  <span>
    Most Northern: 
  </span>
  <span>
    
  </span>
  <br />
  <span>
    Most Southern: 
  </span>
  <span>
    
  </span>
  <br />
  <span>
    Total Population:
  </span>
  <span>
    0
  </span>
  <br />
  <br />
</div>
`;
